Deep signal, deep magic

alxhub opened the pull request.

alxhub committed:
refactor(core): export more from signals package

Add some missing exports that are useful in creating `deepSignal` to the
primitives package.

alxhub committed:
refactor(core): implement `projectedSignal` and `structuralSignal`

These are new signal variants that optimize the way updates are written in
large chains of `projectedSignal`s. This optimization primarily avoids
invalidating `projectedSignal`s that aren't affected by writes, even if they
derive from parents whose values are changed. See the added `README.md` for
details.

sheikalthaf commented on packages/core/src/render3/reactivity/deep_signal/README.md:
```diff
@@ -0,0 +1,81 @@
+effect(() => console.log(user())); // {name: 'Alex'}
+
+// Update the company to 'Waymo':
+model.update(value => {...value, company: 'Waymo'});
```
```suggestion
model.update(value => ({...value, company: 'Waymo'}));
```

rdlabo commented:
Thank you very much for the wonderful pull request. This is the solution [I've been waiting for, and I'm looking forward to it](https://github.com/angular/angular/issues/57955).

The only thing I'm concerned about is the name conflict. One of the names will need to be changed to better match the implementation.

https://github.com/angular/angular/blob/prototype/signal-forms/packages/forms/experimental/src/util/deep_signal.ts

alxhub committed:
refactor(core): implement `upgradeSignalToWritable` utility

This function implements a common pattern/building block seen in other signal
APIs, where a base `computed` or other derived signal is made writable by
patching on a `.set` and `.update` method.

pkozlowski-opensource commented on packages/core/test/render3/projected_signal_spec.ts:
```diff
@@ -0,0 +1,172 @@
+    const first = projectedSignal(parent, 'first', {debugName: 'first'});
+    const last = projectedSignal(parent, 'last', {debugName: 'last'});
+
+    // Keep `last` live, ensuring it gets dirty notifications.
+    act(() => log(last));
```
There is also `log` which creates an effect and makes node live, then `act` flushes effects making everything "clean". At least this is my reading of it...
